src/tinyexpr.cpp

```diff
@@ -22,70 +22,97 @@
  * 3. This notice may not be removed or altered from any source distribution.
  */
 
-// This version has been altered for inclusion in fish.
+// This version has been altered and ported to C++ for inclusion in fish.
 #include "tinyexpr.h"
 #include <stdlib.h>
 #include <math.h>
 #include <string.h>
 #include <stdio.h>
 #include <limits.h>
 
+#include <algorithm>
+#include <iterator>
+#include <utility>
+
+// TODO: It would be nice not to rely on a typedef for this, especially one that can only do functions with two args.
 typedef double (*te_fun2)(double, double);
+typedef double (*te_fun1)(double);
+typedef double (*te_fun0)();
 
 enum {
-    TOK_NULL = TE_FUNCTION0+16, TOK_ERROR, TOK_END, TOK_SEP,
-    TOK_OPEN, TOK_CLOSE, TOK_NUMBER, TOK_VARIABLE, TOK_INFIX
+      TE_CONSTANT = 0,
+      TE_FUNCTION0, TE_FUNCTION1, TE_FUNCTION2, TE_FUNCTION3,
+      TOK_NULL, TOK_ERROR, TOK_END, TOK_SEP,
+      TOK_OPEN, TOK_CLOSE, TOK_NUMBER, TOK_INFIX
 };
 
+int get_arity(const int type) {
+    if (type == TE_FUNCTION3) return 3;
+    if (type == TE_FUNCTION2) return 2;
+    if (type == TE_FUNCTION1) return 1;
+    return 0;
+}
 
-enum {TE_CONSTANT = 1};
-
+typedef struct te_expr {
+    int type;
+    union {double value; const void *function;};
+    // TODO: This void pointer is quite ugly.
+    void *parameters[1];
+} te_expr;
+
+// TODO: Rename since variables have been removed.
+typedef struct te_variable {
+    const char *name;
+    const void *address;
+    int type;
+} te_variable;
 
 typedef struct state {
     const char *start;
     const char *next;
     int type;
-    union {double value; const double *bound; const void *function;};
-    void *context;
+    union {double value; const void *function;};
 
-    const te_variable *lookup;
-    int lookup_len;
     te_error_type_t error;
 } state;
 
+/* Parses the input expression and binds variables. */
+/* Returns NULL on error. */
+te_expr *te_compile(const char *expression, te_error_t *error);
 
-#define TYPE_MASK(TYPE) ((TYPE)&0x0000001F)
+/* Evaluates the expression. */
+double te_eval(const te_expr *n);
 
-#define IS_PURE(TYPE) (((TYPE) & TE_FLAG_PURE) != 0)
-#define IS_FUNCTION(TYPE) (((TYPE) & TE_FUNCTION0) != 0)
-#define ARITY(TYPE) ( ((TYPE) & TE_FUNCTION0) ? ((TYPE) & 0x00000007) : 0 )
-#define NEW_EXPR(type, ...) new_expr((type), (const te_expr*[]){__VA_ARGS__})
+/* Frees the expression. */
+/* This is safe to call on NULL pointers. */
+void te_free(te_expr *n);
+
+// TODO: That move there? Ouch. Replace with a proper class with a constructor.
+#define NEW_EXPR(type, ...) new_expr((type), std::move((const te_expr*[]){__VA_ARGS__}))
```

Comment:
Are you sure this doesn't invoke undefined behavior?
I've been pondering about this interesting issue for a while, out of curiosity, and I kinda doubt now that this code is sound in C++ (unlike C). :thinking: 

If you'd like to remove the `std::move` a C++11 solution (without templates) would be this:
```c++
template<typename... Args>
static te_expr *NEW_EXPR(const int type, Args... args) {
    std::array<const te_expr *, sizeof...(Args)> parameters{{args...}};
    return new_expr(type, parameters.data());
}
```

Reply:
>Are you sure this doesn't invoke undefined behavior?

I don't _think_ so, but I'm not much of a language lawyer.

But the real solution here is to make it an actual class - because `new_expr` is basically just a constructor. I had already done that locally a couple of weeks ago, but didn't push it because it's not worth pushing this PR until 3.0 is done.

Reply:
BTW: Turns out, it's actually relatively easy to do it the C way by using a typedef…
```c++
typedef const te_expr* te_expr_array[];
#define NEW_EXPR(type, ...) new_expr((type), te_expr_array{__VA_ARGS__})
```